### ISSUE TITLE
"Request completionItem/resolve failed."

### ISSUE THREAD

**Issue by garyking** [castwide/solargraph#35]
I'm still getting this sometimes, per #34.

```
[Error - 13:27:15] Request completionItem/resolve failed.
  Message: undefined method `new' for nil:NilClass
  Code: -32603 
```

This happened when I typed `@client.`. Upon the dot, the error popped up. When I deleted this line and re-typed it, the error did not pop up. So it seems like it has a problem getting the data the first time around only.

---

Another:

```
[Error - 13:32:21] Request textDocument/completion failed.
  Message: undefined method `match' for nil:NilClass
  Code: -32603 
```

The above was after: `items.each do |item`.

**Comment by castwide** [comment]
In the last few days I fixed some bugs related to synchronization, where a completion request could start running before a document change was finished processing. Now I'm in the process of improving some other aspects of the internal error handling. I'll have an update to the gem ready this week.

**Comment by garyking** [comment]
After typing: `fo` to try and auto-complete `foo`:

```
[Error - 13:58:57] Request textDocument/completion failed.
  Message: Completion request timed out
  Code: -32603 
```

**Comment by castwide** [comment]
Yeah, it's very likely the process choked on an error. Completion requests time out when it takes longer than one second for the map to become available. Certain errors cause the server to hang for a moment while the exception is getting rescued.

**Comment by castwide** [comment]
This should be fixed in gem v0.18.3.

**Comment by garyking** [comment]
Not fixed in 0.18.3. The other ASCII issue was indeed fixed though.

Here is what I get now:

![vsc-ruby](https://user-images.githubusercontent.com/343940/38636777-332bdc1a-3d97-11e8-82e0-ab02b5530d58.gif)

**Comment by castwide** [comment]
The error demonstrated in the above screenshot is fixed in the current version.
